### PR TITLE
better ux for error handling

### PR DESCRIPTION
**Pull request by eahefnawy**

### CODE REVIEW

File: lib/classes/CLI.js [hunk 1]
```diff
@@ -220,6 +220,39 @@ class CLI {
     console.log('');
   }
 
+  printError(e) {
+    this.asciiGreeting();
```
Comment (pmuens) [inline lib/classes/CLI.js]: Do we really need the greeting? I like it in some scenarios but it's very large and distracts from the actual output IMO.

Comment (eahefnawy) [inline lib/classes/CLI.js]: I removed it from the error stuff, instead I'm greeting on serverless init. You're right, greeting shouldn't have anything to do with errors, specially that errors are thrown mid-way when another greeting has already fired

File: lib/classes/Error.js [hunk 2]
```diff
@@ -1,28 +1,49 @@
...
+  let line = '';
+  while (line.length < 56 - errorType.length) {
+    line = `${line}-`;
+  }
+
+  consoleLog(chalk.yellow(` ${errorType} ${line}`));
+
+  if (e.name !== 'ServerlessError') consoleLog(' ');
+
+  consoleLog(chalk.yellow(`     ${e.message}`));
```
Comment (pmuens) [inline lib/classes/Error.js]: We should add one newline above and one below it so that we can simply throw an error and the format looks always the same.

bildschirmfoto 2016-06-23 um 09 14 00
bildschirmfoto 2016-06-23 um 09 14 07

File: lib/classes/PluginManager.js [hunk 3]
```diff
@@ -23,10 +23,22 @@ class PluginManager {
     this.loadServicePlugins(servicePlugins);
   }
 
+  validateCommands(commandsArray) {
+    // TODO: implement an option to get deeper than one level
+    if (!this.commands[commandsArray[0]]) {
+      const errorMessage =
+      `
```
Comment (pmuens) [inline lib/classes/PluginManager.js]: This newline can be removed once the automatic newline addition is done in the Error class.

File: lib/classes/PluginManager.js [hunk 4]
```diff
@@ -35,12 +47,21 @@ class PluginManager {
 
     forEach(options, (value, key) => {
       if (value.required && (this.options[key] === true || !(this.options[key]))) {
-        throw new this.serverless.classes.Error(`Please provide the "${key}" option`);
+        const errorMessage =
+        `
```
Comment (pmuens) [inline lib/classes/PluginManager.js]: This newline can be removed once the automatic newline addition is done in the Error class.

File: lib/classes/Error.js [hunk 5]
```diff
@@ -1,28 +1,49 @@
...
+
+  if (process.env.SLS_DEBUG) {
+    consoleLog(chalk.yellow('  Stack Trace --------------------------------------------'));
+    consoleLog(' ');
+    consoleLog(e.stack);
+    consoleLog(' ');
+  }
+
+  consoleLog(chalk.yellow('  Get Support --------------------------------------------'));
+  consoleLog(`${chalk.yellow('     Docs:          ')}${chalk.white('https://git.io/voP4S')}`);
```
Comment (pmuens) [inline lib/classes/Error.js]: I like the shortened link but prefer Links with "serverless" in it (like https://github.com/serverless/serverless).

Is there an option to rename the slug?

File: lib/classes/PluginManager.js [hunk 6]
```diff
@@ -61,8 +82,12 @@ class PluginManager {
     });
 
     if (hooks.length === 0) {
-      throw new this.serverless.classes.Error('The command you entered was not found. ' +
-        'Did you spell it correctly?');
+      const errorMessage =
+     `
```
Comment (pmuens) [inline lib/classes/PluginManager.js]: This newline (and also the one below) can be removed once the automatic newline addition is done in the Error class.

File: scripts/eslint.sh [hunk 7]
```diff
@@ -1,20 +0,0 @@
-#!/bin/bash
-
-set -e
-
-mkdir -p tmp
-
-npm run lint | tee tmp/lint-output
-
-problems=`cat tmp/lint-output | grep -oE '[0-9]* problems' | grep -o "[0-9]*"`
-ratchet="100"
```
Comment (pmuens) [inline scripts/eslint.sh]: Not sure if correct but I thought that this is the value Travis uses to mark builds as failed when there are too many linting issues (in this case more than 100).

File: lib/plugins/aws/deploy/tests/deployFunctions.js [hunk 8]
```diff
@@ -164,7 +164,7 @@ describe('deployFunctions', () => {
       serverless.config.servicePath = tmpDirPath;
 
       return awsDeploy.zipFunctions().then(() => {
-        expect(awsDeploy.deployedFunctions[0].zipFileData).to.be.not.empty;
+        expect(typeof awsDeploy.deployedFunctions[0].zipFileData).to.not.equal('undefined');
```
Comment (pmuens) [inline lib/plugins/aws/deploy/tests/deployFunctions.js]: The string `"undefined"` or the type `undefined`?

File: lib/plugins/aws/deploy/tests/deployFunctions.js [hunk 9]
```diff
@@ -195,7 +195,7 @@ describe('deployFunctions', () => {
...
```
Comment (pmuens) [inline lib/plugins/aws/deploy/tests/deployFunctions.js]: The string `"undefined"` or the type `undefined`?

File: lib/classes/Error.js (hunk 2 above)
Comment (eahefnawy) [inline lib/classes/Error.js]: oh the extra space in the first case is because of using string interpolation with the error message. But there's couple of lines of code that handles that issue that you're referencing. However we still need to review error messages to make sure there's no conflicts

File: lib/classes/Error.js (hunk 5 above)
Comment (eahefnawy) [inline lib/classes/Error.js]: nope not with git.io ... git.io doesnt even support non github links 😂 .... so we might have to drop it at some point.

I too like "serverless" links, but I just dislike how long the link is 😛

File: scripts/eslint.sh (hunk 7 above)
Comment (eahefnawy) [inline scripts/eslint.sh]: we now have 0 linting issues ... so now that we removed that file, the build will fail if there's even a single linting issue

File: lib/plugins/aws/deploy/tests/deployFunctions.js (hunk 8 above)
Comment (eahefnawy) [inline lib/plugins/aws/deploy/tests/deployFunctions.js]: `typeof` returns a string, so it's the same thing. This test makes sure that zipFileData is added ... it's exactly like the one u had before, but without the linting issue 😄

File: scripts/eslint.sh (hunk 7 above)
Comment (pmuens) [inline scripts/eslint.sh]: Ah got it! Great. Good job!

File: lib/classes/Error.js (hunk 2 above)
Comment (pmuens) [inline lib/classes/Error.js]: Should be done in a different PR. Here's the corresponding issue: https://github.com/serverless/serverless/issues/1380

File: lib/classes/Error.js (hunk 5 above)
Comment (flomotlik) [inline lib/classes/Error.js]: Imho its important for branding that we have the server less links there directly. Can you change them to the actual links?

**Comment by pmuens** [comment]
GTM from my side once https://github.com/serverless/serverless/pull/1378#discussion_r68186194 is resolved and the build succeeds!

Good job! 🙌  🎉
